src/protocol.rs

```diff
@@ -47,61 +53,25 @@ pub struct Participant {
     /// and presign values. This storage is not responsible for storing
     /// round-specific material.
     main_storage: Storage,
-    /// Participant subprotocol for handling auxinfo messages
-    auxinfo_participant: AuxInfoParticipant,
-    /// Participant subprotocol for handling keygen messages
-    keygen_participant: KeygenParticipant,
-    /// Participant subprotocol for handling presign messages
-    presign_participant: PresignParticipant,
+
+    sid: Identifier,
+    participant: P,
+    input: P::Input,
 }
 
-impl Participant {
+impl<P: PublicProtocolParticipant> Participant<P> {
     /// Initialized the participant from a [ParticipantConfig]
-    #[instrument(err(Debug))]
-    pub fn from_config(config: ParticipantConfig) -> Result<Self> {
+    pub fn from_config(config: &ParticipantConfig, sid: Identifier, input: P::Input) -> Self {
         info!("Initializing participant from config.");
 
-        Ok(Participant {
+        Participant {
             id: config.id,
             other_participant_ids: config.other_ids.clone(),
             main_storage: Storage::new(),
-            auxinfo_participant: AuxInfoParticipant::from_ids(config.id, config.other_ids.clone()),
-            keygen_participant: KeygenParticipant::from_ids(config.id, config.other_ids.clone()),
-            presign_participant: PresignParticipant::from_ids(config.id, config.other_ids),
-        })
-    }
-
-    /// Instantiate a new quorum of participants of a specified size. Random
-    /// identifiers are selected
-    #[instrument(skip_all, err(Debug))]
-    pub fn new_quorum<R: RngCore + CryptoRng>(
```

Comment:
note: I moved this functionality to `ParticipantConfig::random_quorum`, since we want to reuse participant IDs among different `Participant` types. Now a caller can create the random config quorum and instantiate the `Participant`s themselves.